AI detection multitools can show camera static and AI eyes

Cruix opened the pull request:
:cl:
add: AI detection multitools can now be toggled to provide a HUD that shows the exact viewports of AI eyes, and the location of nearby camera blind spots.
/:cl:

Artist's rendering:
![39848978-142cddf8-53d8-11e8-8d00-51dec55605b7](https://user-images.githubusercontent.com/16478175/40881796-5f0ded5c-6685-11e8-8d9c-3afd9e3d12ee.png)

The old color-change functionality of the multitool remains unchanged. The hud is toggled with an action button, because multitools already have an attack_self() that does something. When toggled on, you get the hud as long as the multitool is in any slot on your body.

The only way I could find to reliably show exactly where an AI is looking was to place an image on every turf they are looking at. Because AI eyes move around a lot and I do not want them placing images all over the place when they do not have to, if there is nobody in the world currently using an ai detection multitool, they will not update the images until someone is.

Only AI eyes and multicamera zones currently appear. Advanced camera consoles do not appear on the hud, and no longer trigger the proximity color change. Also fixed multitools not appearing in-hands if their icon state changed.

Example screenshot:
![untitled](https://user-images.githubusercontent.com/16478175/40881788-2b03e868-6685-11e8-824e-96dbe0f9c0c8.png)
The AI eye is the red overlay on the left, a 4x4 multicamera area is the orange overlay on the right, and the transparent static in the center is solid static to the AI.

Cruix committed:
Added a toggle action to the AI detection multitool to show AI camera viewports and camera static
Added a plane and plane master for camera static
Fixed things being added to HUDs sometimes appearing for people they should not appear to, and not being removed
Fixed advanced camera consoles sometimes not showing static

cacogen commented:
fantastic

davethwave commented:
A nice buff that's reasonable.

Cruix committed:
Fixed static staying if the AI detection multitool was deleted while in a person's inventory

Spaceinaba commented:
I am an AI player and I approve this PR.

zxaber commented:
Just to be sure, the static overlay updates with borgs too, right?

YakumoChen commented:
Damn this is some cool spy shit, I can just hear spy music in my head looking at this and trying to sneak around in camera blind spots

Ratvar commented:
Maybe make it into glasses if that's a hud?

Dorsisdwarf commented:
>making the ai detector into a highly-useful tool
m-madness

SpaceManiac commented:
Merge conflict.

Also what's your opinion on action button versus, say, alt-clicking? The item is supposed to be stealthy, after all. Otherwise very useful.

Cruix committed:
Merge branch 'master' into secret_agent_multitool

# Conflicts:
#	code/game/objects/items/devices/multitool.dm

Cruix commented:
The kind of people who will pick up the multitool to see if it has an action are the kind of people who would find out anyway.

Cruix commented:
Making the static unclickable will require potentially doubling the already obscene number of images that the camera chunk system already uses.

Cruix committed:
Fixed multitool update_icon() not behaving as expected
Made Ai detection mulitool huds toggleable while stunned
Fixed multiple equip() calls giving AI detection hud multiple times

MrDoomBringer commented:
>not giving credit for my artists rendering

reeeeeeeeee

Cruix committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into secret_agent_multitool

# Conflicts:
#	code/game/objects/items/devices/multitool.dm
#	code/modules/mob/living/silicon/ai/freelook/cameranet.dm
#	code/modules/mob/living/silicon/ai/freelook/chunk.dm
#	code/modules/mob/living/silicon/ai/freelook/eye.dm

Cruix committed:
Added support for transparent camera static, and made ai detection multitools use it

SpaceManiac commented:
The static is working great in testing! Do you think you'd be able to adapt the red and orange areas to `vis_contents` for the clickthrough as well?

Cruix commented:
I am working on that. I would like to do it slightly differently, because they use of datum huds.

Cruix committed:
Fixed ai detect hud images hiding everything under them from mouse clicks

Cyberboss commented:
Where do we stand on this?

Cruix commented:
Waiting for #38754 to be merged or no longer testmerged.

optimumtact commented:
@Cruix ready?

Cruix commented:
Ready

Cyberboss commented on code/game/objects/items/devices/multitool.dm:
```diff
@@ -98,30 +97,89 @@
+	QDEL_NULL(eye)
 	return ..()
 
+/obj/item/multitool/ai_detect/ui_action_click()
```
We don't proliferate ui_action_click anymore. Use a custom action datum

Cruix replied:
I did use a custom action datum, this is here to suppress the default behavior of ui_action_click().

Cyberboss commented on code/modules/mob/living/silicon/ai/freelook/cameranet.dm:
```diff
@@ -18,15 +18,24 @@ GLOBAL_DATUM_INIT(cameranet, /datum/cameranet, new)
-	// The object used in vis_contents of obscured turfs
-	var/vis_contents
+	// The objects used in vis_contents of obscured turfs
+	var/list/vis_contents
```
can you name this something else so as not to cause confusion please

Cyberboss commented on code/modules/mob/living/silicon/ai/freelook/eye.dm:
```diff
@@ -47,18 +100,18 @@
+	if(ai_detector_visible)
+		var/datum/atom_hud/ai_detector/hud = GLOB.huds[DATA_HUD_AI_DETECT]
+		hud.remove_from_hud(src)
+		QDEL_LIST(hud_list[AI_DETECT_HUD])
```
This will resolve to `hud_list[AI_DETECT_HUD].Cut()` which is actually a : operator. No dice

Cruix replied:
It's pretty silly to have to worry about implicit : operators in the expansion of defines

Cruix commented on code/datums/holocall.dm:
```diff
@@ -60,7 +60,6 @@
 		user.remote_control = null
 
 	if(!QDELETED(eye))
-		eye.RemoveImages()
```
This was replaced by code in the eye object's Destroy()

Cruix committed:
Renamed cameranet.vis_contents
Removed an implicit : operator in the expansion of a define

optimumtact commented:
@Cruix conflicting unfortunately, the rest is fine though (despite the stray map diff failure)

Cruix committed:
Merge branch 'master' of https://github.com/tgstation/-tg-station into secret_agent_multitool

# Conflicts:
#	code/game/machinery/computer/camera_advanced.dm
